Sources/AWSSDKSwift/Services/S3/S3+multipart_API.swift

```diff
@@ -251,3 +212,55 @@ public extension S3 {
         }
     }
 }
+
+
+extension S3 {
+    /// used internally in multipartUpload, loads all the parts once the multipart upload has been initiated
+    func multipartUploadParts(_ input: CreateMultipartUploadRequest, uploadId: String, on eventLoop: EventLoop, inputStream: @escaping () -> EventLoopFuture<Data?>) -> Future<[S3.CompletedPart]> {
+        let promise = eventLoop.makePromise(of: [S3.CompletedPart].self)
+        var completedParts: [S3.CompletedPart] = []
+
+        // function uploading part of a file and queueing up upload of the next part
+        func multipartUploadPart(partNumber: Int, uploadId: String, body: Data? = nil) {
+            // create upload data future, if there is no data to load because this is the first time this is called create a succeeded future
+            let uploadResult : Future<[S3.CompletedPart]>
+            if let body = body {
+                let request = S3.UploadPartRequest(
+                    body: body,
+                    bucket: input.bucket,
+                    contentLength: Int64(body.count),
+                    key: input.key,
+                    partNumber: partNumber,
+                    requestPayer: input.requestPayer,
+                    sSECustomerAlgorithm: input.sSECustomerAlgorithm,
+                    sSECustomerKey: input.sSECustomerKey,
+                    sSECustomerKeyMD5: input.sSECustomerKeyMD5,
+                    uploadId: uploadId
+                )
+                // request upload future
+                uploadResult = self.uploadPart(request).hop(to: eventLoop).map { output -> [S3.CompletedPart] in
+                    let part = S3.CompletedPart(eTag: output.eTag, partNumber: partNumber)
+                    completedParts.append(part)
+                    return completedParts
+                }
+            } else {
+                uploadResult = eventLoop.makeSucceededFuture([])
+            }
+
+            // load data EventLoopFuture
+            let result = inputStream()
+                .and(uploadResult)
+                // upload data
+                .map { (data, parts) -> Void in
+                    guard let data = data, data.count > 0 else { return promise.succeed(parts)}
+                    multipartUploadPart(partNumber: partNumber+1, uploadId: uploadId, body: data)
+            }
+            result.whenFailure { error in
+                promise.fail(error)
+            }
+        }
+
+        multipartUploadPart(partNumber: 1, uploadId: uploadId)
```

Comment:
Is this 1 based upload style requested by AWS? Maybe comment why we use 1 instead of 0?

Reply:
AWS expect a number between 1 and 10000, added comment